### PR TITLE
Add baseHref option

### PR DESCRIPTION
**Pull request by emilbayes**
I'm not sure how this PR sits with the philosophy of bankai being almost zero-conf. I'm also not sure that this fills out all the places that a `<base href="some-url" />` would cover. Lets' discuss!

Ref: #309

### CODE REVIEW

**Comment by YerkoPalma** [comment]
I guess the base directory could be guessed, so we can keep bankai philosophy. For instance,

```bash
$ bankai start app/index
```

should result in a basehref of `app`. Something like this should be enough:

```js
var baseHref = path.dirname(entry)
```

**Comment by goto-bus-stop** [comment]
I'd really love if you could drop a bankai built app into any directory and have it just work, without hardcoding the base path…

I thought we could maybe use relative URLs for scripts. when bankai renders routes it knows if it's on `/` or `/deeply/nested/thing` and could make relative paths accordingly. but you can also make all requests hit index.html for a non server rendered app and then the paths would be wrong, so I guess that's not an option. might be a good area for more exploration though.

(it looks like the client side router doesn't work when deployed to a subdirectory right now either, if you do not use hash routing)

Is there a reason why we can't use `<base href="${opts.baseHref}" />`? of course it would be better to have an option-less thing, but `<base href>` seems easier than adding it manually in all places?

Adding `<base href="someroot">` could be done already by installing `hstream` and adding a documentify transform to package.json:

```json
{
  "documentify": {
    "transform": [
      ["hstream", { "head": { "_appendHtml": "<base href='someroot'>" } }]
    ]
  }
}
```

of course that's not ideal but it might be helpful as a noninvasive way to try out the `base href` way.

@YerkoPalma would that address the case where you have a standalone bankai app but deploy it to `xyz.com/app/` (or eg `tilde.town/~username/`)? If you would just start it as `bankai start index.js` locally in its own git repo.

**Comment by emilbayes** [comment]
Adding a `<base />` tag would be too much magic IMO, as that would impact everywhere that resources are loaded with a leading slash

**Comment by YerkoPalma** [comment]
@goto-bus-stop it works for those and I guess that almost every example, at least in unix systems

image

**Comment by emilbayes** [comment]
Instead of having this strand, I think @YerkoPalma proposal was good and have the default base HREF `.`

File: bin.js [hunk 1]
```diff
@@ -81,6 +82,7 @@ var argv = minimist(process.argv.slice(2), {
   } else if (argv.version) {
     console.log(require('./package.json').version)
   } else if (cmd === 'build') {
+    console.log(argv)
```
Comment (yoshuawuyts) [inline bin.js]: ^

**Comment by goto-bus-stop** [comment]
> Adding a `<base />` tag would be too much magic IMO, as that would impact everywhere that resources are loaded with a leading slash, including dependencies you may not control

@emilbayes that sounds like a really rare case to me for web dependencies, but maybe it is different in electron? I'm not very familiar with it honestly.

I would really rather not introduce new config. https://github.com/choojs/bankai/pull/384 added electron project detection—maybe we can use it to do some magic that works for electron projects as a first step?

**Comment by yoshuawuyts** [comment]
Think this approach is pretty neat. Perhaps we could rename `--base-href` to `--base` in the CLI for ergonomics reasons?

**Comment by yoshuawuyts** [comment]
@goto-bus-stop there was also the case of deploying on subdomains that's related to this:

> Allowing a user to override this would ensure applications can be deployed to a subdirectory on a domain without issue (i.e. http://example.com/mychooapp/is/way/down/here).

Think that's a valid usecase for a lot of people - deployment envs can often be out of control of local devs, and less than ideal. Adding an opt would def solve multiple issues at once for folks heh.

**Comment by goto-bus-stop** [comment]
yeah, i guess we don't have many other options here. @emilbayes I think this approach is good, if you rename it to --base we can get this in.

@yoshuawuyts i believe nanorouter has a similar issue and does not work correctly inside a subdirectory unless you use hash routing only; maybe we should also add a `base` or `root` property to it? I'll check if there's an issue open about that already

**Comment by emilbayes** [comment]
@goto-bus-stop Do you also want the internal option and variable names to be just `base`?

**Comment by goto-bus-stop** [comment]
i think it would make the most sense for them to be the same, yes

**Comment by yoshuawuyts** [comment]
updated to use `'base'`

**Comment by ZhouHansen** [comment]
@emilbayes There's an issue, run:
```
 bankai build index --base boowa
```
generate: 
```
<script src='boowa,boowa/ac78130558410d00/bundle.js'>
```

**Pull request merged**